### PR TITLE
Issue 95 - Add Remote Upstream Repository

### PR DESCRIPTION
**Pull request by roweba**
This implements issue #95 
To test:
1. Click "remote" and enter the url of the remote upstream repository and confirm
2. cd into your project repo and type `git remote -v`
3. The upstream repository should appear in the list

### CODE REVIEW

File: app/misc/git.ts [hunk 1]
```diff
@@ -1317,29 +1317,29 @@ function requestLinkModal() {
...
-        refreshAll(repo);
-      },
-        function (err) {
-          console.log("Waiting for repo to be initialised");
-          displayModal("Please select a valid repository");
-        });
-  } else {
-    displayModal("No Path Found.")
+      repository = repo;
+      var result = Git.Remote.createWithFetchspec(repository, "upstream", upstreamRepoPath, 'remotes/upstream/master');
```
Comment (colinwerner) [inline app/misc/git.ts]: Do we plan on extending this dialog in the future? 'git remote' has a lot more than setting upstream/master.

If no future plans, then perhaps rename this to button to Set Upstream Master? Or Set Upstream?

File: app/components/header/header.component.html [hunk 2]
```diff
@@ -107,6 +107,18 @@
...
             <span class="toolbar-Text">Stash</span>
         </li>
+
+        <!--Add Upstream Remote Respository-->
+        <li class="toolbar-Icons" onclick="addRemoteModal()">
+            <a href="#"><img src="./assets/addremote.png" height="40" width="48" class="add-repository-button" title="Add Remote"></a>
+            <span class="toolbar-Text">Remote</span>
+        </li>
+
+        <!--Sync-->
```
Comment (colinwerner) [inline app/components/header/header.component.html]: I would remove Sync until the impl is complete.

File: app/misc/git.ts (hunk 1 above)
Comment (roweba) [reply app/misc/git.ts]: No this is only going to set the upstream. I can rename it

Comment (roweba) [reply app/misc/git.ts]: As for being able to view/delete the upstream we think we can implement these in the next sprint.

File: app/components/header/header.component.html (hunk 2 above)
Comment (roweba) [reply app/components/header/header.component.html]: @colinwerner I have removed sync and also renamed addRemoteRepo() to setUpstreamRepo()

File: app/misc/git.ts (hunk 1 above)
Comment (colinwerner) [reply app/misc/git.ts]: @roweba please submit a new issue to track these new features.

File: app/components/header/header.component.html [hunk 3]
```diff
@@ -363,8 +369,8 @@ <h4 class="modal-title">Info</h4>
         <input type="text" id="origin-path" style=" width: 554px !important" placeholder="https://github.com/ORIGINAL_OWNER/ORIGINAL_OWNER_REPOSITORY.git">
       </div>
       <div class="modal-footer">
-        <button type="button" class="btn btn-primary" data-dismiss="modal" onClick="fetchFromOrigin()">Confirm</button>
-        <button type="button" class="btn btn-primary" data-dismiss="modal">Close</button>
+        <button type="button" class="btn btn-primary" data-dismiss="modal" onClick="fetchFromOrigin()">test</button>
```
Comment (colinwerner) [inline app/components/header/header.component.html]: What's test for?

File: app/misc/git.ts (hunk 1 above)
Comment (roweba) [reply app/misc/git.ts]: Issue #132 and issue #133 have been created.

File: app/components/header/header.component.html [hunk 4]
```diff
@@ -363,8 +369,8 @@ <h4 class="modal-title">Info</h4>
...
+        <!-- <button type="button" class="btn btn-primary" data-dismiss="modal">Cancel</button> -->
```
Comment (colinwerner) [inline app/components/header/header.component.html]: Is this related to your change?

File: app/misc/git.ts [hunk 5]
```diff
@@ -1542,30 +1542,30 @@ function requestLinkModal() {
...
+}
+
+/**
+ This function is called after the user enters the address of the upstream repo to sync from. It sets the
+ entered address as the upstream repo of the forked repository.
+ */
+function setUpstreamRepo() {
+  let repository;
+  let upstreamRepoPath = document.getElementById("remote-path").value;
+  addCommand("git remote add upstream " + upstreamRepoPath);
```
Comment (colinwerner) [inline app/misc/git.ts]: I would move addCommand into line 1559 so that the command only appears when we actually run it.

File: app/components/header/header.component.html (hunk 4 above)
Comment (roweba) [reply app/components/header/header.component.html]: Yeah I renamed the "close" button but then I just decided to remove it cause there's already an x in the corner. I'll take this out

Comment (roweba) [reply app/components/header/header.component.html]: Sorry this was a mistake! I was testing on the wrong modal, thanks for pointing this out

Comment (roweba) [reply app/components/header/header.component.html]: It looks like this modal is from someone previously attempting to add the sync functionality. I deleted it in the latest commit cause it wasn't being used anywhere.

File: app/components/header/header.component.html (hunk 3 above)
Comment (roweba) [reply app/components/header/header.component.html]: I've added functionality to clear the input

File: app/misc/git.ts [hunk 6]
```diff
@@ -1534,39 +1534,38 @@ function cleanRepo() {
...
-          console.log("Waiting for repo to be initialised");
-          displayModal("Please select a valid repository");
-        });
-  } else {
-    displayModal("No Path Found.")
+      repository = repo;
+      var result = Git.Remote.createWithFetchspec(repository, 'upstream', upstreamRepoPath, '+refs/heads/*:refs/remotes/upstream/*');
+      addCommand("git remote add upstream " + upstreamRepoPath);
+      console.log(result)
+    }, function(err) {
```
Comment (colinwerner) [inline app/misc/git.ts]: If I add throw new Error('test error'); to line 1560 I do not see any error :(. Please fix the error handling to show a dialog if an error occurs.

**Comment by roweba** [comment]
I'm going to close this because this functionality is part of PR #134

**Comment by roweba** [comment]
We have now handled the errors. I have reopened this pull request because it would be better to merge this separately from the sync functionality. @colinwerner please review!

**Comment by roweba** [comment]
I'm making the executive decision to merge this because PR #134 has 2 approvals and any outstanding concerns can be brought up there.
